File: CS/PivotGridDemo.Wpf/ModuleResources/PrefilterViewModel.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using DevExpress.Data.Filtering;
using DevExpress.DemoData;
using DevExpress.DemoData.Models;
using DevExpress.Mvvm.DataAnnotations;

namespace PivotGridDemo {
    public class PrefilterViewModel {
        bool inPrefilterSync;

        public PrefilterViewModel() {
            var dates = SalesPersons.Select(x => x.OrderDate.Value.Date).OrderBy(x => x);
            StartDate = SelectionStart = dates.FirstOrDefault();
            EndDate = SelectionEnd = dates.LastOrDefault();
        }

        public virtual DateTime EndDate { get; set; }
        [BindableProperty(true, OnPropertyChangedMethodName = "UpdateSelection")]
        public virtual CriteriaOperator PrefilterCriteria { get; set; }
        public IList<SalesPerson> SalesPersons { get { return NWindDataProvider.SalesPersons; } }
        [BindableProperty(true, OnPropertyChangedMethodName = "UpdateFilter")]
        public virtual DateTime SelectionEnd { get; set; }
        [BindableProperty(true, OnPropertyChangedMethodName = "UpdateFilter")]
        public virtual DateTime SelectionStart { get; set; }
        public virtual DateTime StartDate { get; set; }

        protected void UpdateFilter() {
            if(inPrefilterSync)
                return;
            string str1 = GetCriteria(SelectionStart, true);
            string str2 = GetCriteria(SelectionEnd, false);
            inPrefilterSync = true;
            if(!string.IsNullOrEmpty(str1) && !string.IsNullOrEmpty(str2))
                PrefilterCriteria = CriteriaOperator.Parse(str1 + " And " + str2);
            else
                PrefilterCriteria = CriteriaOperator.Parse(str1 + str2);
            inPrefilterSync = false;
        }

        protected void UpdateSelection() {
            if (inPrefilterSync)
                return;
            inPrefilterSync = true;
            SelectionStart = GetDateTimeValue(GetBinaryOperatorAgrument(PrefilterCriteria, BinaryOperatorType.GreaterOrEqual), StartDate);
            SelectionEnd = GetDateTimeValue(GetBinaryOperatorAgrument(PrefilterCriteria, BinaryOperatorType.LessOrEqual), EndDate);
            inPrefilterSync = false;
        }

        OperandValue GetBinaryOperatorAgrument(CriteriaOperator criteria, BinaryOperatorType type) {
            GroupOperator group = criteria as GroupOperator;
            if(ReferenceEquals(null, group))
                return null;
            BinaryOperator op = group.Operands.OfType<BinaryOperator>()
                .Where(bin => CheckBinaryOperator(bin, type)).FirstOrDefault();
            return ReferenceEquals(null, op) ? null : op.RightOperand as OperandValue;
        }
        bool CheckBinaryOperator(BinaryOperator binaryOperator, BinaryOperatorType type) {
            return object.Equals(binaryOperator.OperatorType, type) && IsOperandProperty(binaryOperator.LeftOperand, "fieldOrderDate");
        }

        DateTime GetDateTimeValue(OperandValue criteria, DateTime defaultValue) {
            DateTime? dateTime = ReferenceEquals(null, criteria) ? null : criteria.Value as DateTime?;
            return dateTime.HasValue ? dateTime.Value : defaultValue;
        }

        bool IsOperandProperty(CriteriaOperator criteria, string fieldName){
            OperandProperty property = criteria as OperandProperty;
            return !ReferenceEquals(null, property) && property.PropertyName == fieldName;
        }

        string GetCriteria(DateTime date, bool isGreater) {
            return string.Format("{0} {1} #{2}#", "fieldOrderDate", isGreater ? ">=" : "<=",
                Convert.ToString(date, CultureInfo.InvariantCulture));
        }
    }
}

File: VB/PivotGridDemo.Wpf/ModuleResources/PrefilterViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Linq
Imports DevExpress.Data.Filtering
Imports DevExpress.DemoData
Imports DevExpress.DemoData.Models
Imports DevExpress.Mvvm.DataAnnotations

Namespace PivotGridDemo
    Public Class PrefilterViewModel
        Private inPrefilterSync As Boolean

        Public Sub New()
            Dim dates = SalesPersons.Select(Function(x) x.OrderDate.Value.Date).OrderBy(Function(x) x)
            SelectionStart = dates.FirstOrDefault()
            StartDate = SelectionStart
            SelectionEnd = dates.LastOrDefault()
            EndDate = SelectionEnd
        End Sub

        Public Overridable Property EndDate() As Date
        <BindableProperty(True, OnPropertyChangedMethodName := "UpdateSelection")>
        Public Overridable Property PrefilterCriteria() As CriteriaOperator
        Public ReadOnly Property SalesPersons() As IList(Of SalesPerson)
            Get
                Return NWindDataProvider.SalesPersons
            End Get
        End Property
        <BindableProperty(True, OnPropertyChangedMethodName := "UpdateFilter")>
        Public Overridable Property SelectionEnd() As Date
        <BindableProperty(True, OnPropertyChangedMethodName := "UpdateFilter")>
        Public Overridable Property SelectionStart() As Date
        Public Overridable Property StartDate() As Date

        Protected Sub UpdateFilter()
            If inPrefilterSync Then
                Return
            End If
            Dim str1 As String = GetCriteria(SelectionStart, True)
            Dim str2 As String = GetCriteria(SelectionEnd, False)
            inPrefilterSync = True
            If (Not String.IsNullOrEmpty(str1)) AndAlso (Not String.IsNullOrEmpty(str2)) Then
                PrefilterCriteria = CriteriaOperator.Parse(str1 & " And " & str2)
            Else
                PrefilterCriteria = CriteriaOperator.Parse(str1 & str2)
            End If
            inPrefilterSync = False
        End Sub

        Protected Sub UpdateSelection()
            If inPrefilterSync Then
                Return
            End If
            inPrefilterSync = True
            SelectionStart = GetDateTimeValue(GetBinaryOperatorAgrument(PrefilterCriteria, BinaryOperatorType.GreaterOrEqual), StartDate)
            SelectionEnd = GetDateTimeValue(GetBinaryOperatorAgrument(PrefilterCriteria, BinaryOperatorType.LessOrEqual), EndDate)
            inPrefilterSync = False
        End Sub

        Private Function GetBinaryOperatorAgrument(ByVal criteria As CriteriaOperator, ByVal type As BinaryOperatorType) As OperandValue
            Dim group As GroupOperator = TryCast(criteria, GroupOperator)
            If ReferenceEquals(Nothing, group) Then
                Return Nothing
            End If
            Dim op As BinaryOperator = group.Operands.OfType(Of BinaryOperator)().Where(Function(bin) CheckBinaryOperator(bin, type)).FirstOrDefault()
            Return If(ReferenceEquals(Nothing, op), Nothing, TryCast(op.RightOperand, OperandValue))
        End Function
        Private Function CheckBinaryOperator(ByVal binaryOperator As BinaryOperator, ByVal type As BinaryOperatorType) As Boolean
            Return Object.Equals(binaryOperator.OperatorType, type) AndAlso IsOperandProperty(binaryOperator.LeftOperand, "fieldOrderDate")
        End Function

        Private Function GetDateTimeValue(ByVal criteria As OperandValue, ByVal defaultValue As Date) As Date
            Dim dateTime? As Date = If(ReferenceEquals(Nothing, criteria), Nothing, CType(criteria.Value, Date?))
            Return If(dateTime.HasValue, dateTime.Value, defaultValue)
        End Function

        Private Function IsOperandProperty(ByVal criteria As CriteriaOperator, ByVal fieldName As String) As Boolean
            Dim [property] As OperandProperty = TryCast(criteria, OperandProperty)
            Return (Not ReferenceEquals(Nothing, [property])) AndAlso [property].PropertyName = fieldName
        End Function

        Private Function GetCriteria(ByVal [date] As Date, ByVal isGreater As Boolean) As String
            Return String.Format("{0} {1} #{2}#", "fieldOrderDate",If(isGreater, ">=", "<="), Convert.ToString([date], CultureInfo.InvariantCulture))
        End Function
    End Class
End Namespace
